proposals/0000-backpack.rst

```diff
@@ -0,0 +1,1185 @@
+        module Concat
+
+    library stringutils-indef-0.1-xxx where
+        dependency concat-indef-0.1-abcdefg[Str=<Str>]
+        module StringUtils
+
+Here is an informal translation of this AST into command line flags:
+
+For typechecking an uninstantiated mixed library via ``ghc --make``:
+
+1. ``"dependency" UnitId ModuleRenaming`` is translated into
+   the flag ``-unit-id "UnitId ModuleRenaming"`` (the
+   unit identity and module renaming are passed as a single
+   argument with a literal space separating them).
+
+2. ``"module" ModuleName`` is translated into the argument ``ModuleName``,
+   identifying an ``hs`` file in the include path.
+
+3. ``"signature" ModuleName`` is translated into the argument
+   ``ModuleName``, identifying an ``hsig`` file in the include path.
+   Every required signature *must* have an ``hsig`` file (unlike
+   the Cabal syntax, where required signatures can be implicit);
+   it is expected that Cabal generates blank signature files for
+   all inherited signatures.  (This restriction simplifies the
+   compilation model: one source file per compilation product. It may be
+   lifted in the future.)
+
+4. The header of a mixed library ``"library" ComponentId`` is translated into
+   ``-this-unit-id UnitId``, where ``UnitId`` consists of ``ComponentId``
+   and a generalized module substitution ``m=<m>``, for each ``m``
+   in the free module variables of ``dependency`` and each ``m``
+   in ``signature m``.
+
+Thus, these two ASTs would translate into these two command lines::
+
+    ghc -this-unit-id "concat-indef-0.1-abcdefg[Str=<Str>]" \
+        --make Str.hsig Concat.hs
+
+    ghc -this-unit-id "stringutils-indef-0.1-xxx[Str=<Str>]" \
+        -unit-id "concat-indef-0.1-abcdefg[Str=<Str>]" \
+        --make Str.hsig StringUtils.hs
+
+A single module in an uninstantiated mixed library can be
+typechecked with ``ghc -c`` by specifying only that module name.
+(this motivates the "redundant" specification of requirements in
+``this-unit-id``; it's not redundant when compiling with ``ghc -c``!)
+Signatures must always be compiled at once using ``ghc --make`` with
+*only* signatures: this operation is also responsible for determining
+that the dependencies are well-typed.  (TODO: Design a more separate
+compilation interface here.)  To continue our example::
+
+    ghc -this-unit-id "concat-indef-0.1-abcdefg[Str=<Str>]" --make Str.hsig
+    ghc -this-unit-id "concat-indef-0.1-abcdefg[Str=<Str>]" -c Concat.hs
+
+    ghc -this-unit-id "stringutils-indef-0.1-xxx[Str=<Str>]" \
+        -unit-id "concat-indef-0.1-abcdefg[Str=<Str>]" --make Str.hsig
+    ghc -this-unit-id "stringutils-indef-0.1-xxx[Str=<Str>]" \
+        -unit-id "concat-indef-0.1-abcdefg[Str=<Str>]" -c StringUtils.hs
+
+For compiling a mixed library instantiated with module substitution
+``ModuleSubst``, for engineering reasons there are two differences:
+
+    1.  For ``-unit-id``, we specify *hashed* unit identifiers instead
```

Comment:
Engineering reasons should be put in a separate little subsection. What you want to get clear in your mind first is typechecking and compiling. You're assuming the compiling bit, really.